stratum/hal/lib/tdi/es2k/es2k_sde_packetio.cc

```diff
@@ -0,0 +1,428 @@
+  }
+  VLOG(1) << "Received " << buffer.size() << " byte packet from CPU "
+          << StringToHex(buffer);
+
+  return ::util::OkStatus();
+}
+
+void Es2kSdeWrapper::PktIoRxCallback(
+    std::unique_ptr<::tdi::TableKey> key,
+    std::unique_ptr<::tdi::TableData> data,
+    std::unique_ptr<::tdi::NotificationParams> pkts_info_recv, void* cookie) {
+  std::cout << "[Satish] PktIoRxCallback called" << std::endl;
+
+  uint64_t port_id;
+  uint64_t queue_id;
+  uint64_t nb_pkts;
+  std::vector<uint64_t> pkt_len;  /* len of pkts */
+  std::vector<uint64_t> pkt_data; /* pointers to pkts data */
+
+  // parse callback params
+  pkts_info_recv->getValue(PORT_ID, &port_id);
+  pkts_info_recv->getValue(QUEUE_ID, &queue_id);
+  pkts_info_recv->getValue(NB_PKTS, &nb_pkts);
+  pkts_info_recv->getValue(PKT_LEN, &pkt_len);
+  pkts_info_recv->getValue(PKT_DATA, &pkt_data);
+
+  printf("[satish] pktio: %lu pkts received on port_id: %lu, queue_id: %lu\n",
+         nb_pkts, port_id, queue_id);
+
+  Es2kSdeWrapper* tdi_sde_wrapper = Es2kSdeWrapper::GetSingleton();
+
+  const char* raw_data;
+  for (uint64_t i = 0; i < nb_pkts; i++) {
+    printf("pktio: pkt_len: %lu\n", pkt_len[i]);
+    raw_data = reinterpret_cast<const char*>(pkt_data[i]);
+
+    // satish: Only for testing: dump the pkt headers
+    printf("[satish] pktio: pkt received:\n");
+    dump_pkt(raw_data);
+
+    // satish: TODO: Need to find a way to get device_id
+    tdi_sde_wrapper->HandlePacketRx(0, raw_data, pkt_len[i]);
+  }
+}
+
+// StartPacketIo registers callbacks for all the ports and queues in specified
+// in PacketIoConfig
+::util::Status Es2kSdeWrapper::StartPacketIo(int dev_id) {
+  if (pktio_config_.ports_size() == 0) {
+    LOG(INFO) << "pktio not configured";
+    std::cout << "[satish] pktio not configured" << std::endl;
+    return ::util::OkStatus();
+  }
+
+  std::cout << "[satish] pktio configured" << std::endl;
+
+  // create Target
+  const ::tdi::Device* device = nullptr;
+  ::tdi::DevMgr::getInstance().deviceGet(dev_id, &device);
+  std::unique_ptr<::tdi::Target> dev_tgt;
+  device->createTarget(&dev_tgt);
+
+  // get table
+  const ::tdi::Table* table;
+  RETURN_IF_TDI_ERROR(tdi_info_->tableFromNameGet(PKTIO_TABLE_NAME, &table));
+
+  // Allocate rx params
+  std::unique_ptr<::tdi::NotificationParams> rx_notification_params;
+  tdi_status_t status = table->notificationRegistrationParamsAllocate(
+      RX, &rx_notification_params);
+  if (status != BF_SUCCESS) {
+    return MAKE_ERROR(::util::error::Code::INTERNAL)
+           << "rx_notification allocation fail ";
+  }
+
+  // Allocate tx params
+  std::unique_ptr<::tdi::NotificationParams> tx_notification_params;
+  table->notificationRegistrationParamsAllocate(TX, &tx_notification_params);
+  if (status != BF_SUCCESS) {
+    return MAKE_ERROR(::util::error::Code::INTERNAL)
+           << "tx_notification allocation fail ";
+  }
+
+  // Register rx and tx callbacks for all the ports and queues
+  int port_id;
```

Comment:
```suggestion
  int port_id = 0;
```
There's a condition between the definition of port_id and the assignment of its value. port_id should be initialized on definition.

Reply:
Done.